File: todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/tasks/TasksViewModel.java
/*
 * Copyright 2016, The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.example.android.architecture.blueprints.todoapp.tasks;

import android.arch.lifecycle.ViewModel;
import android.support.annotation.NonNull;

import com.example.android.architecture.blueprints.todoapp.data.Task;
import com.example.android.architecture.blueprints.todoapp.mvibase.MviAction;
import com.example.android.architecture.blueprints.todoapp.mvibase.MviIntent;
import com.example.android.architecture.blueprints.todoapp.mvibase.MviResult;
import com.example.android.architecture.blueprints.todoapp.mvibase.MviView;
import com.example.android.architecture.blueprints.todoapp.mvibase.MviViewModel;
import com.example.android.architecture.blueprints.todoapp.mvibase.MviViewState;

import java.util.ArrayList;
import java.util.List;

import io.reactivex.Observable;
import io.reactivex.ObservableTransformer;
import io.reactivex.functions.BiFunction;
import io.reactivex.subjects.PublishSubject;

import static com.example.android.architecture.blueprints.todoapp.util.UiNotificationStatus.SHOW;
import static com.google.common.base.Preconditions.checkNotNull;

/**
 * Listens to user actions from the UI ({@link TasksFragment}), retrieves the data and updates the
 * UI as required.
 */
public class TasksViewModel extends ViewModel implements MviViewModel<TasksIntent, TasksViewState> {

    /**
     * Proxy subject used to keep the stream alive even after the UI gets recycled.
     * This is basically used to keep ongoing events and the last cached State alive
     * while the UI disconnects and reconnects on config changes.
     */
    @NonNull
    private PublishSubject<TasksIntent> mIntentsSubject;
    @NonNull
    private Observable<TasksViewState> mStatesObservable;
    /**
     * Contains and executes the business logic of all emitted actions.
     */
    @NonNull
    private TasksActionProcessorHolder mActionProcessorHolder;

    public TasksViewModel(@NonNull TasksActionProcessorHolder taskActionProcessorHolder) {
        this.mActionProcessorHolder = checkNotNull(taskActionProcessorHolder, "taskActionProcessorHolder cannot be null");

        mIntentsSubject = PublishSubject.create();
        mStatesObservable = compose();
    }

    @Override
    public void processIntents(Observable<TasksIntent> intents) {
        intents.subscribe(mIntentsSubject);
    }

    @Override
    public Observable<TasksViewState> states() {
        return mStatesObservable;
    }

    /**
     * Compose all components to create the stream logic
     */
    private Observable<TasksViewState> compose() {
        return mIntentsSubject
                .compose(intentFilter)
                .map(this::actionFromIntent)
                .compose(mActionProcessorHolder.actionProcessor)
                // Cache each state and pass it to the reducer to create a new state from
                // the previous cached one and the latest Result emitted from the action processor.
                // The Scan operator is used here for the caching.
                .scan(TasksViewState.idle(), reducer)
                // Emit the last one event of the stream on subscription
                // Useful when a View rebinds to the ViewModel after rotation.
                .replay(1)
                // Create the stream on creation without waiting for anyone to subscribe
                // This allows the stream to stay alive even when the UI disconnects and
                // match the stream's lifecycle to the ViewModel's one.
                .autoConnect(0);
    }

    /**
     * take only the first ever InitialIntent and all intents of other types
     * to avoid reloading data on config changes
     */
    private ObservableTransformer<TasksIntent, TasksIntent> intentFilter =
            intents -> intents.publish(shared ->
                    Observable.merge(
                            shared.ofType(TasksIntent.InitialIntent.class).take(1),
                            shared.filter(intent -> !(intent instanceof TasksIntent.InitialIntent))
                    )
            );

    /**
     * Translate an {@link MviIntent} to an {@link MviAction}.
     * Used to decouple the UI and the business logic to allow easy testings and reusability.
     */
    private TasksAction actionFromIntent(MviIntent intent) {
        if (intent instanceof TasksIntent.InitialIntent) {
            return TasksAction.LoadTasks.loadAndFilter(true, TasksFilterType.ALL_TASKS);
        }
        if (intent instanceof TasksIntent.ChangeFilterIntent) {
            return TasksAction.LoadTasks.loadAndFilter(false,
                    ((TasksIntent.ChangeFilterIntent) intent).filterType());
        }
        if (intent instanceof TasksIntent.RefreshIntent) {
            return TasksAction.LoadTasks.load(((TasksIntent.RefreshIntent) intent).forceUpdate());
        }
        if (intent instanceof TasksIntent.ActivateTaskIntent) {
            return TasksAction.ActivateTaskAction.create(
                    ((TasksIntent.ActivateTaskIntent) intent).task());
        }
        if (intent instanceof TasksIntent.CompleteTaskIntent) {
            return TasksAction.CompleteTaskAction.create(
                    ((TasksIntent.CompleteTaskIntent) intent).task());
        }
        if (intent instanceof TasksIntent.ClearCompletedTasksIntent) {
            return TasksAction.ClearCompletedTasksAction.create();
        }
        throw new IllegalArgumentException("do not know how to treat this intent " + intent);
    }

    /**
     * The Reducer is where {@link MviViewState}, that the {@link MviView} will use to
     * render itself, are created.
     * It takes the last cached {@link MviViewState}, the latest {@link MviResult} and
     * creates a new {@link MviViewState} by only updating the related fields.
     * This is basically like a big switch statement of all possible types for the {@link MviResult}
     */
    private static BiFunction<TasksViewState, TasksResult, TasksViewState> reducer =
            (previousState, result) -> {
                TasksViewState.Builder stateBuilder = previousState.buildWith();
                if (result instanceof TasksResult.LoadTasks) {
                    TasksResult.LoadTasks loadResult = (TasksResult.LoadTasks) result;
                    switch (loadResult.status()) {
                        case SUCCESS:
                            TasksFilterType filterType = loadResult.filterType();
                            if (filterType == null) {
                                filterType = previousState.tasksFilterType();
                            }
                            List<Task> tasks = filteredTasks(checkNotNull(loadResult.tasks()), filterType);
                            return stateBuilder.isLoading(false).tasks(tasks).tasksFilterType(filterType).build();
                        case FAILURE:
                            return stateBuilder.isLoading(false).error(loadResult.error()).build();
                        case IN_FLIGHT:
                            return stateBuilder.isLoading(true).build();
                    }
                } else if (result instanceof TasksResult.CompleteTaskResult) {
                    TasksResult.CompleteTaskResult completeTaskResult =
                            (TasksResult.CompleteTaskResult) result;
                    switch (completeTaskResult.status()) {
                        case SUCCESS:
                            stateBuilder.taskComplete(completeTaskResult.uiNotificationStatus() == SHOW);
                            if (completeTaskResult.tasks() != null) {
                                List<Task> tasks =
                                        filteredTasks(checkNotNull(completeTaskResult.tasks()),
                                                previousState.tasksFilterType());
                                stateBuilder.tasks(tasks);
                            }
                            return stateBuilder.build();
                        case FAILURE:
                            return stateBuilder.error(completeTaskResult.error()).build();
                        case IN_FLIGHT:
                            return stateBuilder.build();
                    }
                } else if (result instanceof TasksResult.ActivateTaskResult) {
                    TasksResult.ActivateTaskResult activateTaskResult =
                            (TasksResult.ActivateTaskResult) result;
                    switch (activateTaskResult.status()) {
                        case SUCCESS:
                            stateBuilder.taskActivated(activateTaskResult.uiNotificationStatus() == SHOW);
                            if (activateTaskResult.tasks() != null) {
                                List<Task> tasks =
                                        filteredTasks(checkNotNull(activateTaskResult.tasks()),
                                                previousState.tasksFilterType());
                                stateBuilder.tasks(tasks);
                            }
                            return stateBuilder.build();
                        case FAILURE:
                            return stateBuilder.error(activateTaskResult.error()).build();
                        case IN_FLIGHT:
                            return stateBuilder.build();
                    }
                } else if (result instanceof TasksResult.ClearCompletedTasksResult) {
                    TasksResult.ClearCompletedTasksResult clearCompletedTasks =
                            (TasksResult.ClearCompletedTasksResult) result;
                    switch (clearCompletedTasks.status()) {
                        case SUCCESS:
                            stateBuilder.completedTasksCleared(clearCompletedTasks.uiNotificationStatus() == SHOW);
                            if (clearCompletedTasks.tasks() != null) {
                                List<Task> tasks =
                                        filteredTasks(checkNotNull(clearCompletedTasks.tasks()),
                                                previousState.tasksFilterType());
                                stateBuilder.tasks(tasks);
                            }
                            return stateBuilder.build();
                        case FAILURE:
                            return stateBuilder.error(clearCompletedTasks.error()).build();
                        case IN_FLIGHT:
                            return stateBuilder.build();
                    }
                } else {
                    throw new IllegalArgumentException("Don't know this result " + result);
                }
                throw new IllegalStateException("Mishandled result? Should not happen (as always)");
            };

    private static List<Task> filteredTasks(@NonNull List<Task> tasks,
                                            @NonNull TasksFilterType filterType) {
        List<Task> filteredTasks = new ArrayList<>(tasks.size());
        switch (filterType) {
            case ALL_TASKS:
                filteredTasks.addAll(tasks);
                break;
            case ACTIVE_TASKS:
                for (Task task : tasks) {
                    if (task.isActive()) filteredTasks.add(task);
                }
                break;
            case COMPLETED_TASKS:
                for (Task task : tasks) {
                    if (task.isCompleted()) filteredTasks.add(task);
                }
                break;
        }
        return filteredTasks;
    }
}

File: app/src/main/java/com/example/android/architecture/blueprints/todoapp/tasks/TasksViewModel.kt
/*
 * Copyright 2016, The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.example.android.architecture.blueprints.todoapp.tasks

import android.arch.lifecycle.ViewModel
import com.example.android.architecture.blueprints.todoapp.data.Task
import com.example.android.architecture.blueprints.todoapp.mvibase.MviAction
import com.example.android.architecture.blueprints.todoapp.mvibase.MviIntent
import com.example.android.architecture.blueprints.todoapp.mvibase.MviResult
import com.example.android.architecture.blueprints.todoapp.mvibase.MviView
import com.example.android.architecture.blueprints.todoapp.mvibase.MviViewModel
import com.example.android.architecture.blueprints.todoapp.mvibase.MviViewState
import com.example.android.architecture.blueprints.todoapp.tasks.TasksAction.ActivateTaskAction
import com.example.android.architecture.blueprints.todoapp.tasks.TasksAction.ClearCompletedTasksAction
import com.example.android.architecture.blueprints.todoapp.tasks.TasksAction.CompleteTaskAction
import com.example.android.architecture.blueprints.todoapp.tasks.TasksAction.LoadTasksAction
import com.example.android.architecture.blueprints.todoapp.tasks.TasksFilterType.ACTIVE_TASKS
import com.example.android.architecture.blueprints.todoapp.tasks.TasksFilterType.ALL_TASKS
import com.example.android.architecture.blueprints.todoapp.tasks.TasksFilterType.COMPLETED_TASKS
import com.example.android.architecture.blueprints.todoapp.tasks.TasksResult.ActivateTaskResult
import com.example.android.architecture.blueprints.todoapp.tasks.TasksResult.ClearCompletedTasksResult
import com.example.android.architecture.blueprints.todoapp.tasks.TasksResult.CompleteTaskResult
import com.example.android.architecture.blueprints.todoapp.tasks.TasksResult.LoadTasksResult
import com.example.android.architecture.blueprints.todoapp.util.notOfType
import io.reactivex.Observable
import io.reactivex.ObservableTransformer
import io.reactivex.functions.BiFunction
import io.reactivex.subjects.PublishSubject

/**
 * Listens to user actions from the UI ([TasksFragment]), retrieves the data and updates the
 * UI as required.
 *
 * @property actionProcessorHolder Contains and executes the business logic of all emitted
 * actions.
 */
class TasksViewModel(
    private val actionProcessorHolder: TasksActionProcessorHolder
) : ViewModel(), MviViewModel<TasksIntent, TasksViewState> {

  /**
   * Proxy subject used to keep the stream alive even after the UI gets recycled.
   * This is basically used to keep ongoing events and the last cached State alive
   * while the UI disconnects and reconnects on config changes.
   */
  private val intentsSubject: PublishSubject<TasksIntent> = PublishSubject.create()
  private val statesObservable: Observable<TasksViewState> = compose()

  /**
   * take only the first ever InitialIntent and all intents of other types
   * to avoid reloading data on config changes
   */
  private val intentFilter: ObservableTransformer<TasksIntent, TasksIntent>
    get() = ObservableTransformer { intents ->
      intents.publish { shared ->
        Observable.merge(
            shared.ofType(TasksIntent.InitialIntent::class.java).take(1),
            shared.notOfType(TasksIntent.InitialIntent::class.java)
        )
      }
    }

  override fun processIntents(intents: Observable<TasksIntent>) {
    intents.subscribe(intentsSubject)
  }

  override fun states(): Observable<TasksViewState> = statesObservable

  /**
   * Compose all components to create the stream logic
   */
  private fun compose(): Observable<TasksViewState> {
    return intentsSubject
        .compose(intentFilter)
        .map(this::actionFromIntent)
        .compose(actionProcessorHolder.actionProcessor)
        // Cache each state and pass it to the reducer to create a new state from
        // the previous cached one and the latest Result emitted from the action processor.
        // The Scan operator is used here for the caching.
        .scan(TasksViewState.idle(), reducer)
        // When a reducer just emits previousState, there's no reason to call render. In fact,
        // redrawing the UI in cases like this can cause jank (e.g. messing up snackbar animations
        // by showing the same snackbar twice in rapid succession).
        .distinctUntilChanged()
        // Emit the last one event of the stream on subscription
        // Useful when a View rebinds to the ViewModel after rotation.
        .replay(1)
        // Create the stream on creation without waiting for anyone to subscribe
        // This allows the stream to stay alive even when the UI disconnects and
        // match the stream's lifecycle to the ViewModel's one.
        .autoConnect(0)
  }

  /**
   * Translate an [MviIntent] to an [MviAction].
   * Used to decouple the UI and the business logic to allow easy testings and reusability.
   */
  private fun actionFromIntent(intent: TasksIntent): TasksAction {
    return when (intent) {
      is TasksIntent.InitialIntent -> LoadTasksAction(true, TasksFilterType.ALL_TASKS)
      is TasksIntent.RefreshIntent -> LoadTasksAction(intent.forceUpdate, null)
      is TasksIntent.ActivateTaskIntent -> ActivateTaskAction(intent.task)
      is TasksIntent.CompleteTaskIntent -> CompleteTaskAction(intent.task)
      is TasksIntent.ClearCompletedTasksIntent -> ClearCompletedTasksAction
      is TasksIntent.ChangeFilterIntent -> LoadTasksAction(false, intent.filterType)
    }
  }

  companion object {
    /**
     * The Reducer is where [MviViewState], that the [MviView] will use to
     * render itself, are created.
     * It takes the last cached [MviViewState], the latest [MviResult] and
     * creates a new [MviViewState] by only updating the related fields.
     * This is basically like a big switch statement of all possible types for the [MviResult]
     */
    private val reducer = BiFunction { previousState: TasksViewState, result: TasksResult ->
      when (result) {
        is LoadTasksResult -> when (result) {
          is LoadTasksResult.Success -> {
            val filterType = result.filterType ?: previousState.tasksFilterType
            val tasks = filteredTasks(result.tasks, filterType)
            previousState.copy(
                isLoading = false,
                tasks = tasks,
                tasksFilterType = filterType
            )
          }
          is LoadTasksResult.Failure -> previousState.copy(isLoading = false, error = result.error)
          is LoadTasksResult.InFlight -> previousState.copy(isLoading = true)
        }
        is CompleteTaskResult -> when (result) {
          is CompleteTaskResult.Success ->
            previousState.copy(
                taskComplete = true,
                tasks = filteredTasks(result.tasks, previousState.tasksFilterType)
            )
          is CompleteTaskResult.Failure -> previousState.copy(error = result.error)
          is CompleteTaskResult.InFlight -> previousState
          is CompleteTaskResult.HideUiNotification -> previousState.copy(taskComplete = false)
        }
        is ActivateTaskResult -> when (result) {
          is ActivateTaskResult.Success ->
            previousState.copy(
                taskActivated = true,
                tasks = filteredTasks(result.tasks, previousState.tasksFilterType)
            )
          is ActivateTaskResult.Failure -> previousState.copy(error = result.error)
          is ActivateTaskResult.InFlight -> previousState
          is ActivateTaskResult.HideUiNotification -> previousState.copy(taskActivated = false)
        }
        is ClearCompletedTasksResult -> when (result) {
          is ClearCompletedTasksResult.Success ->
            previousState.copy(
                completedTasksCleared = true,
                tasks = filteredTasks(result.tasks, previousState.tasksFilterType)
            )
          is ClearCompletedTasksResult.Failure -> previousState.copy(error = result.error)
          is ClearCompletedTasksResult.InFlight -> previousState
          is ClearCompletedTasksResult.HideUiNotification ->
            previousState.copy(completedTasksCleared = false)
        }
      }
    }

    private fun filteredTasks(tasks: List<Task>, filterType: TasksFilterType): List<Task> {
      return when (filterType) {
        ALL_TASKS -> tasks
        ACTIVE_TASKS -> tasks.filter(Task::active)
        COMPLETED_TASKS -> tasks.filter(Task::completed)
      }
    }
  }
}
